### ISSUE TITLE
Always send nil values in the response payload

### ISSUE THREAD

**Issue by vsgoulart** [camunda/camunda#23607]
## Description

_A clear and concise description of what this proposal is about and what problem it solves._

Standardize JSON responses by always including nil properties as explicit `null` values or empty arrays, rather than omitting them. This change aims to improve API consistency and ease of use for consumers.

The motivation for this is to simplify the validation of these payloads. For example, in TypeScript if you had a field that has 2 possible types and it's nullable the types would be translate to something like `string | number | undefined`, which makes it necessary for the consumer to do extra checks for these optional properties.

Example of an existing public API that follows the same guideline: [Stripe API](https://docs.stripe.com/api/customers/object#customer_object-phone)

## Proposal

Always send props with nil values.

Current:
```json
{
  "mandatoryProp": 1
  // missing optional fields
}
```

Proposal:
```json
{
 "mandatoryProp": 1,
  "optionalArray": [],
  "optionalProp": null
}
```

## Decisions

_Will be filled in by the project DRIs after a peer review phase._

**Comment by nathansandi** [comment]
My Thoughts on Nullability @tmetzke

I prefer to stick with the current approach and continue using Optional instead of null for the following reasons:

	1.	Ambiguity: A null object can create confusion. Consumers of the API may not be able to distinguish whether null represents an error, missing data, or something else, leading to misinterpretation.
	2.	Error-Prone: Handling null improperly can easily lead to NullPointerExceptions or similar issues in client applications, which could result in crashes or unexpected behavior.
	3.	Industry Standards: Many API best practices and specifications (like REST and GraphQL) advise against returning null values. Instead, they recommend offering clear, well-defined responses. For example, Google’s JSON style guide suggests avoiding null where possible. That said, this might be somewhat **subjective**, and I acknowledge there are different perspectives on this. So related to Industry standards, this depends. (For example the counter example on Stripe API)

As mentioned on Slack, this approach has its pros and cons, but overall, I believe it ensures better clarity and error-handling.

**Comment by vsgoulart** [comment]
@nathansandi Regarding point 2, wouldn't trying to access a property that's not there also cause a `NullPointerException`?

**Comment by aleksander-dytko** [comment]
I don't have any customer insights indicating that they struggle with the current (V1) approach. This means that they're probably fine with whatever we build as long as we document it and describe it in the migration guide. 

Due to the consistency with V1, I would propose sticking with the current approach. 

@tmetzke what's your opinion on this?

**Comment by tmetzke** [comment]
* There is no right or wrong here, as far as I can see, both options have their pros and cons.
  * I agree that having a stable API in terms of attributes returned in responses is more convenient to handle in some languages than leaving out null-value attributes. In some languages, it doesn't make much of a difference (e.g. in Java it is a null check either way with most frameworks that do the (de-)serialization).
  * Sending less data is also preferred in many cases which counts toward the current approach.
  * V1 consistency also counts toward the current approach.
  * We have to check if sending nil attributes causes any unexpected regressions in our codebase. It might not but we have to check.
* The proposed approach can still be taken later if there is a high demand for it, as far as I can see.
* I would not make this a priority for now.
* I would also not dismiss this as invalid but simply as not having enough priority for now.

@vsgoulart, let us know if you think this blocks your work or makes our API tremendously worse to use and should definitely have a higher priority. As mentioned, we would also need to check that this does not introduce any regressions and we need to communicate a change in behavior regarding V1 API.

**Comment by vsgoulart** [comment]
@tmetzke Yeah, this doesn't block me. So feel free to keep as is in the current V2 API, it's just more convenient to handle responses in JS this way (I believe in Elixir and Go also have the same benefit)

> Due to the consistency with V1, I would propose sticking with the current approach.

At least on Tasklist (and on Operate until I left the team) the approach we used was to always send the properties as `null`

**Comment by tmetzke** [comment]
Valid point, thanks, @vsgoulart 👍 
I only looked at Operate API, to be honest, to assess what we do in V1 and wrongfully assumed that it's aligned.

### Local test

In a local `8.6.2` C8Run instance, Operate and Tasklist V1 APIs behave differently.

Operate V1 API only returns attributes that are set.
Calling `http://localhost:8080/v1/process-instances/2251799813686106` yields
```
{
  "key": 2251799813686106,
  "processVersion": 1,
  "bpmnProcessId": "send-email",
  "startDate": "2024-10-24T08:44:25.384+0000",
  "state": "ACTIVE",
  "incident": false,
  "processDefinitionKey": 2251799813686105,
  "tenantId": "\u003Cdefault\u003E"
}
```
The [endpoint specification](https://docs.camunda.io/docs/next/apis-tools/operate-api/specifications/by-key-1/) features more attributes though, like `parentKey`, `parentFlowNodeInstanceKey`, `endDate`, and `parentProcessInstanceKey`.

If I start a process instance as a call activity, I can see some of those attributes returned, for example:
```
{
  "key": 2251799813695217,
  "processVersion": 1,
  "bpmnProcessId": "send-email",
  "parentKey": 2251799813695212,
  "parentFlowNodeInstanceKey": 2251799813695216,
  "startDate": "2024-10-24T09:38:00.890+0000",
  "state": "ACTIVE",
  "incident": false,
  "processDefinitionKey": 2251799813686105,
  "tenantId": "\u003Cdefault\u003E"
}
```

Tasklist V1 API also returns nil attributes.
Calling `http://localhost:8080/v1/tasks/2251799813686111` yields:
```
{
  "id": "2251799813686111",
  "name": "Enter Message",
  "taskDefinitionId": "enterMessageTask",
  "processName": "Send Email",
  "creationDate": "2024-10-24T10:44:25.384+0200",
  "completionDate": null,
  "assignee": null,
  "taskState": "CREATED",
  "formKey": null,
  "formId": null,
  "formVersion": null,
  "isFormEmbedded": false,
  "processDefinitionKey": "2251799813686105",
  "processInstanceKey": "2251799813686106",
  "tenantId": "\u003Cdefault\u003E",
  "dueDate": null,
  "followUpDate": null,
  "candidateGroups": null,
  "candidateUsers": null,
  "implementation": "JOB_WORKER",
  "priority": 50
}
```

**Comment by jonathanlukas** [comment]
I would vote for always sending all null values explicitly.

An option would be to control the behaviour with a flag `returnNullValues` (name tbd) that is `false` by default.

**Comment by tmetzke** [comment]
This will be implemented with https://github.com/camunda/camunda/issues/46175
